### ISSUE TITLE
Potential STATUS_HEAP_CORRUPTION from dsp.get_info() implementation

### ISSUE THREAD

**Issue by GitGhillie** [lebedec/libfmod#19]
Hi!

In my crate I loop through all dsp's (including plugin DSPs) in all channel groups and call `get_info()` on each. It looks like sometimes this causes a STATUS_HEAP_CORRUPTION when my game starts. Sadly I'm not able to reproduce the issue consistently.

However, looking into [`get_info()`](https://github.com/lebedec/libfmod/blob/2.02.22/libfmod/src/lib.rs#L7894), it looks like there might be a safety issue. `name`, which should be a 32 byte buffer, is initialized like this: 
`let name = CString::from_vec_unchecked(b"".to_vec()).into_raw();`
and after FMOD_DSP_GetInfo cast into a String like this:
```
CString::from_raw(name)
                        .into_string()
                        .map_err(Error::String)?
```

The comments on CString::from_raw state the following: "It should be noted that the length isn't just "recomputed," but that the recomputed length must match the original length from the CString::into_raw call. This means the CString::into_raw/from_raw methods should not be used when passing the string to C functions that can modify the string's length."

So I tried initializing it with the same length, 32:
`let name = CString::from_vec_unchecked(b"foodfoodfoodfoodfoodfoodfoodfood".to_vec()).into_raw();`

And since then I haven't seen the issue (but then again, I wasn't able to reproduce consistently so I could be on a lucky streak so to say).
What do you think? Could this be an issue or would you like to have something reproducible?

**Comment by BrianWiz** [comment]
Its worth noting that this seems to happen on Windows. Let me know if you need something reproducible.

**Comment by lebedec** [comment]
Hi! Thanks for investigation. I'll try to fix this issue.

**Comment by lebedec** [comment]
So, there are 20 incorrect function implementations that have out argument with buffer-based string type:

```
OUT FMOD_ChannelGroup_GetName name
OUT FMOD_DSP_GetParameterFloat valuestr
OUT FMOD_DSP_GetParameterInt valuestr
OUT FMOD_DSP_GetParameterBool valuestr
OUT FMOD_DSP_GetParameterData valuestr
OUT FMOD_DSP_GetInfo name
OUT FMOD_Sound_GetName name 
OUT FMOD_Sound_GetSyncPointInfo name
OUT FMOD_SoundGroup_GetName name
OUT FMOD_Studio_Bank_GetStringInfo path
OUT FMOD_Studio_CommandReplay_GetCommandString buffer
OUT FMOD_Studio_EventDescription_GetParameterLabelByIndex label
OUT FMOD_Studio_EventDescription_GetParameterLabelByName label
OUT FMOD_Studio_EventDescription_GetParameterLabelByID label
OUT FMOD_Studio_System_GetParameterLabelByName label
OUT FMOD_Studio_System_GetParameterLabelByID label
OUT FMOD_System_GetDriverInfo name
OUT FMOD_System_GetPluginInfo name
OUT FMOD_System_GetRecordDriverInfo name
OUT FMOD_System_GetNetworkProxy proxy
````

The problem is specification of buffer length varies in each function:
- an explicit length argument
- almost explicit part of the structure definition (like Dsp name)
- implicit text notes in the documentation 😢 

The solution is to write correct string retrieving manually for each function via `[0;len].as_mut_ptr() + CStr::from_ptr`.

But the number of manual overrites is too big for the main concept of `libfmod` like "automatically generated library" =) Need more time to think how to do it better.

Workaround for now if it blocks your work: 

```rust
pub fn get_dsp_info(dsp: &Dsp) -> Result<(String, u32, i32, i32, i32), Error> {
    unsafe {
        let mut name = [0i8; 32];
        let mut version = u32::default();
        let mut channels = i32::default();
        let mut configwidth = i32::default();
        let mut configheight = i32::default();
        match ffi::FMOD_DSP_GetInfo(
            dsp.as_mut_ptr(),
            name.as_mut_ptr(),
            &mut version,
            &mut channels,
            &mut configwidth,
            &mut configheight,
        ) {
            ffi::FMOD_OK => Ok((
                CStr::from_ptr(name.as_ptr()).to_string_lossy().to_string(),
                version,
                channels,
                configwidth,
                configheight,
            )),
            error => Err(err_fmod!("FMOD_DSP_GetInfo", error)),
        }
    }
}
```

**Comment by lebedec** [comment]
Fixed in `libfmod = "2.222.5"`

**Comment by GitGhillie** [comment]
Awesome, thanks again for the quick resolution! Things are looking solid now
